File: BoopGame.py
from dataclasses import dataclass, field
from abc import ABC, abstractmethod
from typing import Optional, List, Tuple
from enum import Enum

class PieceType(Enum):
    GATITO = 1
    GATO = 2

@dataclass
class Player:
    name: str
    color: str  # 'orange' o 'gray'
    gatitos_disponibles: int = 8
    gatos_disponibles: int = 0
    
    def can_place_gatito(self) -> bool:
        return self.gatitos_disponibles > 0
    
    def can_place_gato(self) -> bool:
        return self.gatos_disponibles > 0
    
    def use_gatito(self):
        if self.can_place_gatito():
            self.gatitos_disponibles -= 1
    
    def use_gato(self):
        if self.can_place_gato():
            self.gatos_disponibles -= 1
    
    def return_gatito(self):
        self.gatitos_disponibles += 1
    
    def return_gato(self):
        self.gatos_disponibles += 1
    
    def promote_gatitos_to_gato(self, count: int = 3):
        """Convierte 3 gatitos en 1 gato"""

        self.gatitos_disponibles += count
        self.gatos_disponibles += 3

@dataclass
class Piece(ABC):
    position: Tuple[int, int]
    player: Player
    piece_type: Optional[PieceType] = None
    
    @property
    @abstractmethod
    def weight(self) -> int:
        pass
    


@dataclass
class Gatito(Piece):
    def __post_init__(self):
        self.piece_type = PieceType.GATITO
    
    @property
    def weight(self) -> int:
        return 1
    
    @property
    def can_boop(self) -> bool:
        return True

@dataclass
class Gato(Piece):
    def __post_init__(self):
        self.piece_type = PieceType.GATO
    
    @property
    def weight(self) -> int:
        return 2
    
    @property
    def can_boop(self) -> bool:
        return False  # Los gatos adultos no pueden ser empujados


@dataclass
class Board:
    size: int = 6
    grid: List[List[Optional[Piece]]] = field(default_factory=lambda: [[None for _ in range(6)] for _ in range(6)])
    
    def is_valid_position(self, row: int, col: int) -> bool:
        return 0 <= row < self.size and 0 <= col < self.size
    
    def is_empty(self, row: int, col: int) -> bool:
        return self.is_valid_position(row, col) and self.grid[row][col] is None
    
    def place_piece(self, piece: Piece, row: int, col: int) -> bool:
        if self.is_empty(row, col):
            self.grid[row][col] = piece
            piece.position = (row, col)
            return True
        return False
    
    def remove_piece(self, row: int, col: int) -> Optional[Piece]:
        if self.is_valid_position(row, col):
            piece = self.grid[row][col]
            self.grid[row][col] = None
            return piece
        return None
    
    def get_piece(self, row: int, col: int) -> Optional[Piece]:
        if self.is_valid_position(row, col):
            return self.grid[row][col]
        return None
    
    def get_adjacent_positions(self, row: int, col: int) -> List[Tuple[int, int]]:
        """Obtiene todas las posiciones adyacentes (8 direcciones)"""
        directions = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]
        adjacent = []
        for dr, dc in directions:
            new_row, new_col = row + dr, col + dc
            if self.is_valid_position(new_row, new_col):
                adjacent.append((new_row, new_col))
        return adjacent
    
    def boop_pieces(self, placed_row: int, placed_col: int, placed_piece: Piece) -> List[Piece]:
        """Empuja las piezas adyacentes y retorna las que salen del tablero"""
        booped_out = []
        adjacent_positions = self.get_adjacent_positions(placed_row, placed_col)
        
        for adj_row, adj_col in adjacent_positions:
            piece = self.get_piece(adj_row, adj_col)
            if piece and piece.can_boop:
                # Calcular dirección del empuje
                dr = adj_row - placed_row
                dc = adj_col - placed_col
                new_row = adj_row + dr
                new_col = adj_col + dc
                
                # Verificar si la pieza puede moverse a la nueva posición
                if self.is_valid_position(new_row, new_col) and self.is_empty(new_row, new_col):
                    # Remover la pieza de su posición actual
                    self.remove_piece(adj_row, adj_col)
                    # Colocarla en la nueva posición
                    self.place_piece(piece, new_row, new_col)
                elif not self.is_valid_position(new_row, new_col):
                    # La pieza sale del tablero
                    self.remove_piece(adj_row, adj_col)
                    booped_out.append(piece)
                # Si la nueva posición está ocupada, la pieza se queda estática (no se mueve)
        
        return booped_out
    
    def find_lines_of_three(self, player: Player) -> List[List[Tuple[int, int]]]:
        """Encuentra todas las líneas de 3 piezas del mismo jugador"""
        lines = []
        directions = [(0, 1), (1, 0), (1, 1), (1, -1)]  # horizontal, vertical, diagonal
        
        for row in range(self.size):
            for col in range(self.size):
                piece = self.get_piece(row, col)
                if piece and piece.player == player:
                    for dr, dc in directions:
                        line = [(row, col)]
                        for i in range(1, 3):
                            new_row = row + i * dr
                            new_col = col + i * dc
                            next_piece = self.get_piece(new_row, new_col)
                            if (next_piece and next_piece.player == player and 
                                type(next_piece) == type(piece)):
                                line.append((new_row, new_col))
                            else:
                                break
                        
                        if len(line) == 3:
                            lines.append(line)
        
        return lines
    
    def display(self):
        """Muestra el tablero en la consola"""
        print("  ", end="")
        for col in range(self.size):
            print(f" {col} ", end="")
        print()
        
        for row in range(self.size):
            print(f"{row} ", end="")
            for col in range(self.size):
                piece = self.get_piece(row, col)
                if piece is None:
                    print(" . ", end="")
                elif isinstance(piece, Gatito):
                    symbol = "o" if piece.player.color == "orange" else "x"
                    print(f" {symbol} ", end="")
                elif isinstance(piece, Gato):
                    symbol = "O" if piece.player.color == "orange" else "X"
                    print(f" {symbol} ", end="")
            print()


@dataclass
class Boop:
    board: Board = field(default_factory=Board)
    player1: Player = field(default_factory=lambda: Player("Jugador 1", "orange"))
    player2: Player = field(default_factory=lambda: Player("Jugador 2", "gray"))
    current_player: Optional[Player] = None
    game_over: bool = False
    winner: Optional[Player] = None
    
    def __post_init__(self):
        self.current_player = self.player1
    
    def switch_player(self):
        self.current_player = self.player2 if self.current_player == self.player1 else self.player1
    
    def place_piece(self, row: int, col: int, piece_type: PieceType) -> bool:
        """Coloca una pieza en el tablero y ejecuta la lógica del juego"""
        if self.game_over:
            print("El juego ha terminado!")
            return False
        
        if not self.board.is_empty(row, col):
            print("Esa posición ya está ocupada!")
            return False
        
        # Verificar si el jugador puede colocar el tipo de pieza solicitado
        if piece_type == PieceType.GATITO and not self.current_player.can_place_gatito():
            print("No tienes más gatitos disponibles!")
            return False
        elif piece_type == PieceType.GATO and not self.current_player.can_place_gato():
            print("No tienes gatos disponibles!")
            return False
        
        # Crear la pieza
        if piece_type == PieceType.GATITO:
            piece = Gatito((row, col), self.current_player)
            self.current_player.use_gatito()
        else:
            piece = Gato((row, col), self.current_player)
            self.current_player.use_gato()
        
        # Colocar la pieza en el tablero
        self.board.place_piece(piece, row, col)
        
        # Ejecutar el boop (empujar piezas adyacentes)
        booped_out = self.board.boop_pieces(row, col, piece)
        
        # Devolver las piezas que salieron del tablero a sus jugadores
        for booped_piece in booped_out:
            if isinstance(booped_piece, Gatito):
                booped_piece.player.return_gatito()
            else:
                booped_piece.player.return_gato()
        
        # Verificar líneas de 3 gatitos para promoción
        self.check_and_promote_gatitos()
        
        # Verificar condiciones de victoria
        self.check_victory()
        
        if not self.game_over:
            self.switch_player()
        
        return True
    
    def check_and_promote_gatitos(self):
        """Verifica y promueve líneas de 3 gatitos a gatos adultos"""
        for player in [self.player1, self.player2]:
            lines = self.board.find_lines_of_three(player)
            
            for line in lines:
                # Verificar si todos son gatitos
                pieces = [self.board.get_piece(row, col) for row, col in line]
                if all(isinstance(piece, Gatito) for piece in pieces):
                    # Remover los gatitos del tablero
                    for row, col in line:
                        self.board.remove_piece(row, col)
                    
                    # Promover gatitos a gato
                    player.promote_gatitos_to_gato(3)
                    
                    print(f"¡{player.name} ha graduado a 3 gaatitos!")
    
    def check_victory(self):
        """Verifica las condiciones de victoria"""
        for player in [self.player1, self.player2]:
            lines = self.board.find_lines_of_three(player)
            
            for line in lines:
                # Verificar si todos son gatos adultos
                pieces = [self.board.get_piece(row, col) for row, col in line]
                if all(isinstance(piece, Gato) for piece in pieces):
                    self.game_over = True
                    self.winner = player
                    print(f"¡{player.name} ha ganado con 3 gatos adultos en línea!")
                    return
    
    def display_game_state(self):
        """Muestra el estado actual del juego"""
        print("\n" + "="*30)
        print(f"Turno de: {self.current_player.name} ({self.current_player.color})")
        print(f"Gatitos disponibles: {self.current_player.gatitos_disponibles}")
        print(f"Gatos disponibles: {self.current_player.gatos_disponibles}")
        print("="*30)
        self.board.display()
        print("="*30)
    
    def get_player_input(self) -> Tuple[int, int, PieceType]:
        """Obtiene la entrada del jugador"""
        while True:
            try:
                print(f"\n{self.current_player.name}, es tu turno!")
                print("Ingresa tu movimiento en formato: fila,columna,tipo")
                print("Tipo: 'g' para gatito, 'G' para gato adulto")
                print("Ejemplo: 2,3,g")
                
                input_str = input("Tu movimiento: ").strip()
                parts = input_str.split(',')
                
                if len(parts) != 3:
                    print("Formato incorrecto. Usa: fila,columna,tipo")
                    continue
                
                row = int(parts[0])
                col = int(parts[1])
                piece_char = parts[2].strip()
                
                if piece_char == 'g':
                    piece_type = PieceType.GATITO
                elif piece_char == 'G':
                    piece_type = PieceType.GATO
                else:
                    print("Tipo de pieza inválido. Usa 'g' para gatito o 'G' para gato adulto")
                    continue
                
                return row, col, piece_type
                
            except (ValueError, IndexError):
                print("Entrada inválida. Intenta de nuevo.")
    
    def play(self):
        """Bucle principal del juego"""
        print("¡Bienvenido al juego Boop!")
        print("Objetivo: Forma una línea de 3 gatos adultos para ganar")
        print("Los gatitos en línea de 3 se convierten en gatos adultos")
        print("Los gatitos empujan a piezas adyacentes (boop!)")
        print("Los gatos adultos no pueden ser empujados")
        
        while not self.game_over:
            self.display_game_state()
            row, col, piece_type = self.get_player_input()
            
            if not self.place_piece(row, col, piece_type):
                print("Movimiento inválido. Intenta de nuevo.")
                continue
        
        print("\n" + "="*40)
        print("¡JUEGO TERMINADO!")
        if self.winner:
            print(f"¡Felicitaciones {self.winner.name}! ¡Has ganado!")
        print("="*40)
        self.board.display()


# Función para iniciar el juego
def main():
    game = Boop()
    game.play()

if __name__ == "__main__":
    main()

File: BoopGame.cpp
#include <iostream>
#include <vector>
#include <string>
#include <memory>
#include <utility>
#include <sstream>
#include <algorithm>

using namespace std;

// Forward declarations
class Player;
class Piece;
class Gatito;
class Gato;

// Enumeración para tipos de pieza
enum class PieceType {
    GATITO = 1,
    GATO = 2
};

// Clase Player
class Player {
public:
    string name;
    string color;
    int gatitos_disponibles;
    int gatos_disponibles;

    Player(const string& n, const string& c) 
        : name(n), color(c), gatitos_disponibles(8), gatos_disponibles(0) {}

    bool canPlaceGatito() const {
        return gatitos_disponibles > 0;
    }

    bool canPlaceGato() const {
        return gatos_disponibles > 0;
    }

    void useGatito() {
        if (canPlaceGatito()) {
            gatitos_disponibles--;
        }
    }

    void useGato() {
        if (canPlaceGato()) {
            gatos_disponibles--;
        }
    }

    void returnGatito() {
        gatitos_disponibles++;
    }

    void returnGato() {
        gatos_disponibles++;
    }

    void promoteGatitosToGato(int count = 3) {
        gatitos_disponibles += count;
        gatos_disponibles += 3;
    }
};

// Clase base abstracta Piece
class Piece {
public:
    pair<int, int> position;
    Player* player;
    PieceType piece_type;

    Piece(pair<int, int> pos, Player* p, PieceType type) 
        : position(pos), player(p), piece_type(type) {}

    virtual ~Piece() = default;

    virtual int weight() const = 0;
    virtual bool canBoop() const = 0;
    virtual char getSymbol() const = 0;
};

// Clase Gatito
class Gatito : public Piece {
public:
    Gatito(pair<int, int> pos, Player* p) 
        : Piece(pos, p, PieceType::GATITO) {}

    int weight() const override {
        return 1;
    }

    bool canBoop() const override {
        return true;
    }

    char getSymbol() const override {
        return player->color == "orange" ? 'o' : 'x';
    }
};

// Clase Gato
class Gato : public Piece {
public:
    Gato(pair<int, int> pos, Player* p) 
        : Piece(pos, p, PieceType::GATO) {}

    int weight() const override {
        return 2;
    }

    bool canBoop() const override {
        return true;
    }

    char getSymbol() const override {
        return player->color == "orange" ? 'O' : 'X';
    }
};

// Clase Board
class Board {
public:
    int size;
    vector<vector<Piece*>> grid;

    Board(int s = 6) : size(s) {
        grid.resize(size, vector<Piece*>(size, nullptr));
    }

    ~Board() {
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                delete grid[i][j];
            }
        }
    }

    bool isValidPosition(int row, int col) const {
        return row >= 0 && row < size && col >= 0 && col < size;
    }

    bool isEmpty(int row, int col) const {
        return isValidPosition(row, col) && grid[row][col] == nullptr;
    }

    bool placePiece(Piece* piece, int row, int col) {
        if (isEmpty(row, col)) {
            grid[row][col] = piece;
            piece->position = {row, col};
            return true;
        }
        return false;
    }

    Piece* removePiece(int row, int col) {
        if (isValidPosition(row, col)) {
            Piece* piece = grid[row][col];
            grid[row][col] = nullptr;
            return piece;
        }
        return nullptr;
    }

    Piece* getPiece(int row, int col) const {
        if (isValidPosition(row, col)) {
            return grid[row][col];
        }
        return nullptr;
    }

    vector<pair<int, int>> getAdjacentPositions(int row, int col) const {
        vector<pair<int, int>> adjacent;
        int directions[8][2] = {{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}};
        
        for (int i = 0; i < 8; i++) {
            int newRow = row + directions[i][0];
            int newCol = col + directions[i][1];
            if (isValidPosition(newRow, newCol)) {
                adjacent.push_back({newRow, newCol});
            }
        }
        return adjacent;
    }

    vector<Piece*> boopPieces(int placedRow, int placedCol, Piece* placedPiece) {
        vector<Piece*> boopedOut;
        vector<pair<int, int>> adjacentPositions = getAdjacentPositions(placedRow, placedCol);

        for (const auto& pos : adjacentPositions) {
            int adjRow = pos.first;
            int adjCol = pos.second;
            Piece* piece = getPiece(adjRow, adjCol);
            
            if (piece && piece->canBoop() && placedPiece->weight() >= piece->weight()) {
                int dr = adjRow - placedRow;
                int dc = adjCol - placedCol;
                int newRow = adjRow + dr;
                int newCol = adjCol + dc;

                if (isValidPosition(newRow, newCol) && isEmpty(newRow, newCol)) {
                    removePiece(adjRow, adjCol);
                    placePiece(piece, newRow, newCol);
                } else if (!isValidPosition(newRow, newCol)) {
                    removePiece(adjRow, adjCol);
                    boopedOut.push_back(piece);
                }
            }
        }
        return boopedOut;
    }

    vector<vector<pair<int, int>>> findLinesOfThree(Player* player) const {
        vector<vector<pair<int, int>>> lines;
        int directions[4][2] = {{0, 1}, {1, 0}, {1, 1}, {1, -1}};

        for (int row = 0; row < size; row++) {
            for (int col = 0; col < size; col++) {
                Piece* piece = getPiece(row, col);
                if (piece && piece->player == player) {
                    for (int d = 0; d < 4; d++) {
                        vector<pair<int, int>> line = {{row, col}};
                        int dr = directions[d][0];
                        int dc = directions[d][1];

                        for (int i = 1; i < 3; i++) {
                            int newRow = row + i * dr;
                            int newCol = col + i * dc;
                            Piece* nextPiece = getPiece(newRow, newCol);
                            
                            if (nextPiece && nextPiece->player == player &&
                                nextPiece->piece_type == piece->piece_type) {
                                line.push_back({newRow, newCol});
                            } else {
                                break;
                            }
                        }

                        if (line.size() == 3) {
                            lines.push_back(line);
                        }
                    }
                }
            }
        }
        return lines;
    }

    void display() const {
        cout << "  ";
        for (int col = 0; col < size; col++) {
            cout << " " << col << " ";
        }
        cout << endl;

        for (int row = 0; row < size; row++) {
            cout << row << " ";
            for (int col = 0; col < size; col++) {
                Piece* piece = getPiece(row, col);
                if (piece == nullptr) {
                    cout << " . ";
                } else {
                    cout << " " << piece->getSymbol() << " ";
                }
            }
            cout << endl;
        }
    }
};

// Clase principal del juego Boop
class Boop {
public:
    Board board;
    Player player1;
    Player player2;
    Player* currentPlayer;
    bool gameOver;
    Player* winner;

    Boop() : player1("Jugador 1", "orange"), player2("Jugador 2", "gray"),
             currentPlayer(&player1), gameOver(false), winner(nullptr) {}

    void switchPlayer() {
        currentPlayer = (currentPlayer == &player1) ? &player2 : &player1;
    }

    bool placePiece(int row, int col, PieceType pieceType) {
        if (gameOver) {
            cout << "El juego ha terminado!" << endl;
            return false;
        }

        if (!board.isEmpty(row, col)) {
            cout << "Esa posición ya está ocupada!" << endl;
            return false;
        }

        if (pieceType == PieceType::GATITO && !currentPlayer->canPlaceGatito()) {
            cout << "No tienes más gatitos disponibles!" << endl;
            return false;
        } else if (pieceType == PieceType::GATO && !currentPlayer->canPlaceGato()) {
            cout << "No tienes gatos disponibles!" << endl;
            return false;
        }

        Piece* piece;
        if (pieceType == PieceType::GATITO) {
            piece = new Gatito({row, col}, currentPlayer);
            currentPlayer->useGatito();
        } else {
            piece = new Gato({row, col}, currentPlayer);
            currentPlayer->useGato();
        }

        board.placePiece(piece, row, col);

        vector<Piece*> boopedOut = board.boopPieces(row, col, piece);

        for (Piece* boopedPiece : boopedOut) {
            if (dynamic_cast<Gatito*>(boopedPiece)) {
                boopedPiece->player->returnGatito();
            } else {
                boopedPiece->player->returnGato();
            }
            delete boopedPiece;
        }

        checkAndPromoteGatitos();
        checkVictory();

        if (!gameOver) {
            switchPlayer();
        }

        return true;
    }

    void checkAndPromoteGatitos() {
        Player* players[] = {&player1, &player2};
        
        for (Player* player : players) {
            vector<vector<pair<int, int>>> lines = board.findLinesOfThree(player);
            
            for (const auto& line : lines) {
                bool allGatitos = true;
                for (const auto& pos : line) {
                    Piece* piece = board.getPiece(pos.first, pos.second);
                    if (!dynamic_cast<Gatito*>(piece)) {
                        allGatitos = false;
                        break;
                    }
                }

                if (allGatitos) {
                    for (const auto& pos : line) {
                        Piece* piece = board.removePiece(pos.first, pos.second);
                        delete piece;
                    }
                    player->promoteGatitosToGato(3);
                    cout << "¡" << player->name << " ha graduado a 3 gatitos!" << endl;
                }
            }
        }
    }

    void checkVictory() {
        Player* players[] = {&player1, &player2};
        
        for (Player* player : players) {
            vector<vector<pair<int, int>>> lines = board.findLinesOfThree(player);
            
            for (const auto& line : lines) {
                bool allGatos = true;
                for (const auto& pos : line) {
                    Piece* piece = board.getPiece(pos.first, pos.second);
                    if (!dynamic_cast<Gato*>(piece)) {
                        allGatos = false;
                        break;
                    }
                }

                if (allGatos) {
                    gameOver = true;
                    winner = player;
                    cout << "¡" << player->name << " ha ganado con 3 gatos adultos en línea!" << endl;
                    return;
                }
            }
        }
    }

    void displayGameState() const {
        cout << "\n==============================" << endl;
        cout << "Turno de: " << currentPlayer->name << " (" << currentPlayer->color << ")" << endl;
        cout << "Gatitos disponibles: " << currentPlayer->gatitos_disponibles << endl;
        cout << "Gatos disponibles: " << currentPlayer->gatos_disponibles << endl;
        cout << "==============================" << endl;
        board.display();
        cout << "==============================" << endl;
    }

    tuple<int, int, PieceType> getPlayerInput() {
        while (true) {
            cout << "\n" << currentPlayer->name << ", es tu turno!" << endl;
            cout << "Ingresa tu movimiento en formato: fila,columna,tipo" << endl;
            cout << "Tipo: 'g' para gatito, 'G' para gato adulto" << endl;
            cout << "Ejemplo: 2,3,g" << endl;
            cout << "Tu movimiento: ";

            string input;
            getline(cin, input);

            stringstream ss(input);
            string rowStr, colStr, typeStr;

            if (getline(ss, rowStr, ',') && getline(ss, colStr, ',') && getline(ss, typeStr)) {
                try {
                    int row = stoi(rowStr);
                    int col = stoi(colStr);
                    
                    // Eliminar espacios
                    typeStr.erase(remove_if(typeStr.begin(), typeStr.end(), ::isspace), typeStr.end());

                    PieceType pieceType;
                    if (typeStr == "g") {
                        pieceType = PieceType::GATITO;
                    } else if (typeStr == "G") {
                        pieceType = PieceType::GATO;
                    } else {
                        cout << "Tipo de pieza inválido. Usa 'g' para gatito o 'G' para gato adulto" << endl;
                        continue;
                    }

                    return {row, col, pieceType};
                } catch (...) {
                    cout << "Entrada inválida. Intenta de nuevo." << endl;
                }
            } else {
                cout << "Formato incorrecto. Usa: fila,columna,tipo" << endl;
            }
        }
    }

    void play() {
        cout << "¡Bienvenido al juego Boop!" << endl;
        cout << "Objetivo: Forma una línea de 3 gatos adultos para ganar" << endl;
        cout << "Los gatitos en línea de 3 se convierten en gatos adultos" << endl;
        cout << "Los gatitos empujan a piezas adyacentes (boop!)" << endl;
        cout << "Los gatos adultos no pueden ser empujados" << endl;

        while (!gameOver) {
            displayGameState();
            auto [row, col, pieceType] = getPlayerInput();

            if (!placePiece(row, col, pieceType)) {
                cout << "Movimiento inválido. Intenta de nuevo." << endl;
            }
        }

        cout << "\n========================================" << endl;
        cout << "¡JUEGO TERMINADO!" << endl;
        if (winner) {
            cout << "¡Felicitaciones " << winner->name << "! ¡Has ganado!" << endl;
        }
        cout << "========================================" << endl;
        board.display();
    }
};

int main() {
    Boop game;
    game.play();
    return 0;
}